core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java

```diff
@@ -217,7 +217,11 @@ private boolean isAliasPresent(Iterable<String> stageOldNameAliasPair) {
 
   void validateJoinKeySchemas(Map<String, Schema> inputSchemas, Map<String, List<String>> joinKeys,
                               FailureCollector collector) {
-    if (!containsMacro(JoinerConfig.JOIN_KEYS) && inputSchemas.size() != 0 && joinKeys.size() != inputSchemas.size()) {
+    // Skip validation if joinKeys is a macro, or if any input's output schema is a macro.
+    if (!containsMacro(JoinerConfig.JOIN_KEYS) &&
+      // TODO: Remove isEmpty() check when CDAP-16351 is fixed
+      !inputSchemas.isEmpty() && !inputSchemas.values().stream().anyMatch(v -> v == null) &&
```

Comment:
looks like duplicate code, can it be refactored so it's not in multiple places?

Reply:
Refactored this into `JoinerConfig.inputSchemasAvailable()`